src/databricks/labs/ucx/azure/resources.py

```diff
@@ -121,44 +117,142 @@ def inner() -> dict[str, str]:
+            query = {"api-version": api_version}
+        return self.api_client.do("GET", path, query, headers)
+
+    def put(self, path: str, api_version: str | None = None, body: dict | None = None):
+        headers = {"Content-Type": "application/json"}
+        query: dict[str, str] = {}
+        if api_version is not None:
+            query = {"api-version": api_version}
+        if body is not None:
+            return self.api_client.do("PUT", path, query, headers, body)
+        return None
+
+    def post(self, path: str, body: dict | None = None):
+        headers = {"Content-Type": "application/json"}
+        query: dict[str, str] = {}
+        if body is not None:
+            return self.api_client.do("POST", path, query, headers, body)
+        return self.api_client.do("POST", path, query, headers)
+
+    def delete(self, path: str):
+        # this method is added only to be used in int test to delete the application once tests pass
+        headers = {"Content-Type": "application/json"}
+        query: dict[str, str] = {}
+        return self.api_client.do("DELETE", path, query, headers)
+
+    def token(self):
+        return self._token_source.token()
+
+
+class AzureResources:
+    def __init__(self, azure_mgmt: AzureAPIClient, azure_graph: AzureAPIClient, include_subscriptions=None):
+        if not include_subscriptions:
+            include_subscriptions = []
+        self._azure_mgmt = azure_mgmt
+        self._azure_graph = azure_graph
+        self._include_subscriptions = include_subscriptions
+        self._role_definitions = {}  # type: dict[str, str]
+        self._principals: dict[str, Principal | None] = {}
+
     def _get_subscriptions(self) -> Iterable[AzureSubscription]:
-        for subscription in self._get_resource("/subscriptions", api_version="2022-12-01").get("value", []):
+        for subscription in self._azure_mgmt.get("/subscriptions", "2022-12-01").get("value", []):
             yield AzureSubscription(
                 name=subscription["displayName"],
                 subscription_id=subscription["subscriptionId"],
                 tenant_id=subscription["tenantId"],
             )
 
-    def _tenant_id(self):
-        token = self._token_source.token()
+    def create_service_principal(self, display_name: str) -> PrincipalSecret:
+        try:
+            application_info: dict[str, str] = self._azure_graph.post(
+                "/v1.0/applications", {"displayName": display_name}
+            )
+            app_id = application_info.get("appId")
+            assert app_id is not None
+            service_principal_info: dict[str, str] = self._azure_graph.post(
+                "/v1.0/servicePrincipals", {"appId": app_id}
+            )
+            object_id = service_principal_info.get("id")
+            assert object_id is not None
+            secret_info: dict[str, str] = self._azure_graph.post(f"/v1.0/servicePrincipals/{object_id}/addPassword")
+
+        except PermissionDenied:
```

Comment:
which API call is it permission denied for? :) e.g. what happens to what you've created in `POST /v1.0/applications`? will it be dangling? do you have cleanup?

Reply:
If the user doesnt have permission to create application then it would throw a Permission denied error. If the user can create an application then should also create the serviceprincipal using the application id. both should work or both should fail

Reply:
why it's not in the code comments? ;)